### DISCUSSION TITLE
bucket4j mysql integration

### DISCUSSION THREAD

**Discussion by arminsls** [vladimir-bukhtoyarov/bucket4j#414] [Q&A]
Hello. I decided to keep max limit buckets for every user in mysql database. For example every user is permitted to create only 10 channels in my chat app from the beginning of registration till death!! if it is his first request for creating channel, a bucket with 10 token is created and a token is consumed. and if it is not the first time, bucket table is checked and a token is consumed from the remaining tokens. there could be five or six or more maxLimits for each user(for examle: channel creaqtion max limit, group creation max limit, ...). I want you to suggest me the best way! my users can exceed 10 million and if each one of them user every request with max limit it can reach a 50 or 60 million row table. in your opinion which one is better? insert all buckets for all the max limits in a table or a table for each max limit? and if I you select multi table, I should insert tables settings to sqlproxy configuration at the beginning of app runtime?
I appreciate your help.
@vladimir-bukhtoyarov

**Comment by vladimir-bukhtoyarov** [comment]
> For example every user is permitted to create only 10 channels in my chat app from the beginning of registration till death!!

I do not think that such requirements are related to reate-limiting domain. It looks like regular businness restrictions that can be covered by triggers and similar database provided integrity checks. You should use Bucket4j if and only if your usecase particulary covered by [Token bucket](https://en.wikipedia.org/wiki/Token_bucket), otherwise you should choose another way.

**Reply by arminsls** [reply]
Thanks for the response. I just mentioned the requirement to enlighten the problem. To be straight, the question is, is it best practice to create five or six tables to keep every specific bucket for users in it, or it is better to create just one table and insert all the buckets for all the users in it? Can bucket4j handle reading and inserting data in such tables with more than 50 or 60 million rows, or the performance or speed decreases? And should I use a list of "MySQLSelectForUpdateBasedProxyManager" to build a "sqlProxyConfiguration" with a list of tableSettings? Do you recommend this?

**Reply by vladimir-bukhtoyarov** [reply]
I have no user feedbacks about question that you asked. It is better to do perfomance testing by yourself.
